foundry.md

```diff
@@ -185,24 +187,60 @@ First we have some helpers in K which can:
+Finally, the rule for `#return_foundry` is used to end the execution of the `CALL` OpCode.
+
+```{.k .bytes}
+    rule [foundry.call]:
+         <k> CALL _ CHEAT_ADDR _VALUE ARGSTART ARGWIDTH RETSTART RETWIDTH
+          => #call_foundry #asWord(#range(LM, ARGSTART, 4)) #range(LM, ARGSTART +Int 4, ARGWIDTH -Int 4)
+          ~> #return_foundry RETSTART RETWIDTH
+         ...
+         </k>
+         <localMem> LM </localMem>
+         <output> _ => .ByteArray </output>
+    requires CHEAT_ADDR ==Int #address(FoundryCheat)
+    [priority(40)]
+```
+
+We define two productions named `#return_foundry` and `#call_foundry`, which will be used by each cheat code.
+The rule `foundry.return` will rewrite the `#return_foundry` production into other productions that will place the output of the execution into the local memory, refund the gas value of the call and push the value `1` on the call stack.
+
+```{.k .bytes}
+    syntax KItem ::= "#error_foundry" [klabel(foundry_error)]
+                   | "#return_foundry" Int Int [klabel(foundry_return)]
+                   | "#call_foundry" Int ByteArray [klabel(foundry_call)]
+ // ---------------------------------------------------------------------
+    rule [foundry.return]:
+         <k> #return_foundry RETSTART RETWIDTH
+          => #setLocalMem RETSTART RETWIDTH OUT
+          ~> #refund GCALL
+          ~> 1 ~> #push
+          ... </k>
+         <output> OUT </output>
+         <callGas> GCALL </callGas>
+```
+
+Below, we define rules for the `#call_foundry` production, handling the cheat codes.
+
 #### `assume` - Insert a condition
 
 ```
 function assume(bool) external;
 ```
 
-`call.assume` will match when the `assume` function is called at the [Foundry cheatcode address](https://book.getfoundry.sh/cheatcodes/#cheatcodes-reference).
-This rule then takes a `bool` condition using `#range(LM, ARGSTART +Int 4, 32)` and injects it into the execution path using the `#assume` production.
+`foundry.call.assume` will match when the `assume` cheat code function is called.
+This rule then takes a `bool` condition from the function call data, represented using the `ARGS` variable, and injects it into the execution path using the `#assume` production.
 `==K #bufStrict(32, 1)` is used to mark that the condition should hold.
-```k
 
-    rule [call.assume]:
-         <k> CALL _ CHEAT_ADDR _VALUE ARGSTART _ARGWIDTH _RETSTART _RETWIDTH => #assume(#range(LM, ARGSTART +Int 4, 32) ==K #bufStrict(32, 1)) ~> #finishCheatCode ... </k>
-         <output> _ => .ByteArray </output>
-         <localMem> LM </localMem>
-      requires CHEAT_ADDR ==Int #address(FoundryCheat)
-       andBool #asWord(#range(LM, ARGSTART, 4)) ==Int 1281615202 // selector ( "assume(bool)" )
-      [priority(40)]
+```{.k .bytes}
+    rule [foundry.call.assume]:
+         <k> #call_foundry SELECTOR ARGS => #assume(ARGS ==K #bufStrict(32, 1)) ... </k>
+      requires SELECTOR ==Int selector ( "assume(bool)" )
```

Comment:
This doesn't put the same width restriction on `ARGS` (by doing `#range(ARGS, 0, 32)` basically) as the original rule.

I wonder if we should do that? In the below rule for `deal` you _do_ include that extra restriction.

Reply:
See this issue: https://github.com/runtimeverification/evm-semantics/issues/1410

Reply:
In the `rule` for `deal`, I do that because `ARGS` is supposed to contain two arguments. Here I assume that the restriction is made on line 200 with `#range(LM, ARGSTART +Int 4, ARGWIDTH -Int 4)`.